sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c

```diff
@@ -0,0 +1,349 @@
+                return;
+            }
+            if (!gnrc_lorawan_mac_acquire(mac)) {
+                mlme_confirm->status = -EBUSY;
+                return;
+            }
+
+            if (mac->mlme.backoff_budget < 0) {
+                mlme_confirm->status = -EDQUOT;
+                break;
+            }
+            memcpy(mac->appskey, mlme_request->join.appkey, LORAMAC_APPKEY_LEN);
+            mlme_confirm->status = gnrc_lorawan_send_join_request(mac, mlme_request->join.deveui,
+                                                                  mlme_request->join.appeui, mlme_request->join.appkey, mlme_request->join.dr);
+            break;
+        case MLME_LINK_CHECK:
+            mac->mlme.pending_mlme_opts |= GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_DEFERRED;
+            break;
+        case MLME_SET:
+            _mlme_set(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_GET:
+            _mlme_get(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_RESET:
+            gnrc_lorawan_reset(mac);
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+            break;
+        default:
+            break;
+    }
+}
+
+int _fopts_mlme_link_check_req(lorawan_buffer_t *buf)
+{
+    if (buf) {
+        assert(buf->index + GNRC_LORAWAN_CID_SIZE <= buf->size);
+        buf->data[buf->index++] = GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS;
+    }
+
+    return GNRC_LORAWAN_CID_SIZE;
+}
+
+static int _mlme_link_check_ans(gnrc_lorawan_t *mac, lorawan_buffer_t *fopt)
+{
+    if (fopt->index + GNRC_LORAWAN_FOPT_LINK_ANS_SIZE > fopt->size) {
+        return -EINVAL;
+    }
+    fopt->index++;
+
+    mlme_confirm_t *mlme_confirm = _mlme_allocate(mac);
+    mlme_confirm->link_req.margin = fopt->data[fopt->index++];
+    mlme_confirm->link_req.num_gateways = fopt->data[fopt->index++];
+
+    mlme_confirm->type = MLME_LINK_CHECK;
+    mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+    mac->netdev.event_callback(&mac->netdev, NETDEV_EVENT_MLME_CONFIRM);
+
+    mac->mlme.pending_mlme_opts &= ~GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+
+    return 0;
+}
+
+void gnrc_lorawan_process_fopts(gnrc_lorawan_t *mac, uint8_t *fopts, size_t size)
+{
+    if (!fopts || !size) {
+        return;
+    }
+
+    lorawan_buffer_t buf;
+    _buffer_reset(&buf, fopts, size);
+
+    while (buf.index < buf.size) {
+        switch (*(buf.data)) {
+            case GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS:
+                if (_mlme_link_check_ans(mac, &buf) < 0) {
+                    return;
+                }
+                break;
+            default:
+                return;
+        }
+    }
+}
+
+uint8_t gnrc_lorawan_build_options(gnrc_lorawan_t *mac, lorawan_buffer_t *buf)
+{
+    size_t size = 0;
+
+    size += (mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ) ?
+            _fopts_mlme_link_check_req(buf) : 0;
```

Comment:
though I like ternary, I have the feeling that in this case the following is more readable

```
if (mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ) {
    size += _fopts_mlme_link_check_req(buf);
}
```

Reply:
if you want to optimise than refine whole function to:

```
uint8_t gnrc_lorawan_build_options(gnrc_lorawan_t *mac, lorawan_buffer_t *buf)
{
    return (uint8_t)(mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ) ?
           _fopts_mlme_link_check_req(buf) : 0;
}
```

Reply:
I intentionally left this "size" variable there because there will be more fopts. But I will change it to the `if` pattern

Reply:
this an all comments above: addressed in https://github.com/RIOT-OS/RIOT/pull/11022/commits/a278c06444f9a157fb32daf6740a5837550f35d7